cirq/experiments/random_quantum_circuit.py

```diff
@@ -0,0 +1,229 @@
+from typing import (Callable, Container, Dict, Iterable, List, Sequence,
+                    TYPE_CHECKING, Tuple)
+
+from cirq import circuits, devices, google, ops, value
+from cirq._doc import document
+
+if TYPE_CHECKING:
+    import numpy as np
+    import cirq
+
+# A set of 2-qubit interactions that can be performed simultaneously on a grid.
+TwoQubitInteractionLayer = Container[
+    Tuple[devices.GridQubit, devices.GridQubit]]
+
+
+class GmonLayer(TwoQubitInteractionLayer):
+    """A Gmon-style layer of 2-qubit gates where neighbors don't need to idle.
+
+    These layers contain gates involving all qubits in the lattice. They come
+    in two variants, unstaggered:
+
+    *-* *-* *-*
+    *-* *-* *-*
+    *-* *-* *-*
+    *-* *-* *-*
+    *-* *-* *-*
+    *-* *-* *-*
+
+    and staggered:
+
+    *-* *-* *-*
+    * *-* *-* *
+    *-* *-* *-*
+    * *-* *-* *
+    *-* *-* *-*
+    * *-* *-* *
+
+    Other variants are obtained by offsetting these lattices to the right by
+    some number of columns, and/or transposing into the vertical orientation.
+    There are a total of 4 staggered and 4 unstaggered variants, and a pattern
+    will cycle through one of these groups of 4 to perform all 2-qubit gates.
+
+    The 2x2 unit cells for the unstaggered and staggered versions of this layer
+    are, respectively:
+
+    *-*
+    *-*
+
+    and
+
+    *-*
+    * *-
+
+    with left/top qubits at (0, 0) and (1, 0) in the unstaggered case, or
+    (0, 0) and (1, 1) in the staggered case. Other variants have the same unit
+    cells after transposing and offsetting.
+    """
+
+    def __init__(self,
+                 col_offset: int = 0,
+                 vertical: bool = False,
+                 stagger: bool = False) -> None:
+        """
+        Args:
+            col_offset: Number of columns by which to shift the basic lattice.
+            vertical: Whether gates should be oriented vertically rather than
+                horizontally.
+            stagger: Whether to stagger gates in neighboring rows.
+        """
+        self.col_offset = col_offset
+        self.vertical = vertical
+        self.stagger = stagger
+
+    def __contains__(self, pair) -> bool:
+        """Checks whether a pair is in this layer."""
+        if self.vertical:
+            # Transpose row, col coords for vertical orientation.
+            a, b = pair
+            pair = devices.GridQubit(a.col,
+                                     a.row), devices.GridQubit(b.col, b.row)
+
+        a, b = sorted(pair)
+
+        # qubits should be 1 column apart.
+        if (a.row != b.row) or (b.col != a.col + 1):
+            return False
+
+        # mod to get the position in the 2 x 2 unit cell with column offset.
+        pos = a.row % 2, (a.col - self.col_offset) % 2
+        return pos == (0, 0) or pos == (1, self.stagger)
+
+
+GMON_HARD_PATTERN = (
```

Comment:
nit: Perhaps `GRID_STAGGERED_PATTERN` as it's more descriptive?

Reply:
Done.